src/metabase/models/card.clj

```diff
@@ -140,13 +142,39 @@
       :mbql/query (-> query lib/normalize lib.util/source-card-id)
       nil)))
 
+(defn- card->integer-table-ids
+  "Return integer source table ids for card's :dataset_query."
+  [card]
+  (when-some [query (-> card :dataset_query :query)]
+    (not-empty (filter pos-int? (lib.util/collect-source-tables query)))))
+
+(defn- prefetch-tables-for-cards!
+  "Collect tables from `dataset-cards` and prefetch metadata. Should be used only with metdata provider caching
+  enabled, as per https://github.com/metabase/metabase/pull/45050. Returns `nil`."
+  [dataset-cards]
+  (let [db-id->table-ids (-> (group-by :database_id dataset-cards)
+                             (update-vals (partial into #{} (comp (mapcat card->integer-table-ids)
+                                                                  (remove nil?)))))
+        db-id->mp (into {}
+                        (map (juxt identity lib.metadata.jvm/application-database-metadata-provider))
+                        (keys db-id->table-ids))]
+    (doseq [[db-id table-ids] db-id->table-ids
+            :when (seq table-ids)]
+      (lib.metadata.protocols/metadatas (db-id->mp db-id) :metadata/table table-ids))))
+
 (defn with-can-run-adhoc-query
   "Adds can_run_adhoc_query to each card."
   [cards]
   (let [dataset-cards (filter (comp seq :dataset_query) cards)
         source-card-ids (into #{}
                               (keep (comp source-card-id :dataset_query))
                               dataset-cards)]
+    ;; Prefetching code should not propagate any exception.
+    (try
+      (when lib.metadata.jvm/*metadata-provider-cache*
+        (prefetch-tables-for-cards! dataset-cards))
+      (catch Throwable _
+        (log/errorf "Error prefething cards `%s`." (pr-str (map :id dataset-cards)))))
```

Comment:
Don't throw away the cause! You can attach the `Throwable` as the first argument to `log/errorf` and we'll log a lot more context.

Reply:
I believe `(log/errorf t ...)` arity caputures that. Thanks.